### ISSUE TITLE
Aborted PC2 will always fail with bad descriptor until PC1 regenerated

### ISSUE THREAD

**Issue by karalabe** [filecoin-project/rust-fil-proofs#1327]
I've seen this enough times to decide it's a bug.

Every now and again I need to restart my miner/worker (config changes, lotus update, etc). Since I don't want to wait for 12+ hours for all the seal tasks to finish, I usually just tell it to shut down gracefully.

When I start the workers back up, it picks up where it left off (obviously restarting the PC1 or PC2 or C2, can't continue). However, whilst restarted PC1 or C2 will succeed, restarted PC2 will always fail with "Bad File Descriptor"

```
Oct 24 11:00:38 filecoin lotus-worker[20925]: thread '<unnamed>' panicked at 'failed to flatten and write store: Bad file descriptor (os error 9)', /root/.cargo/registry/src/github.com>
Oct 24 11:00:38 filecoin lotus-worker[20925]: stack backtrace:
Oct 24 11:00:38 filecoin lotus-worker[20925]:    0: backtrace::backtrace::libunwind::trace
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at ./cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/libunwind.rs:86
Oct 24 11:00:38 filecoin lotus-worker[20925]:    1: backtrace::backtrace::trace_unsynchronized
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at ./cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/mod.rs:66
Oct 24 11:00:38 filecoin lotus-worker[20925]:    2: std::sys_common::backtrace::_print_fmt
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libstd/sys_common/backtrace.rs:78
Oct 24 11:00:38 filecoin lotus-worker[20925]:    3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libstd/sys_common/backtrace.rs:59
Oct 24 11:00:38 filecoin lotus-worker[20925]:    4: core::fmt::write
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libcore/fmt/mod.rs:1063
Oct 24 11:00:38 filecoin lotus-worker[20925]:    5: std::io::Write::write_fmt
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libstd/io/mod.rs:1426
Oct 24 11:00:38 filecoin lotus-worker[20925]:    6: std::sys_common::backtrace::_print
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libstd/sys_common/backtrace.rs:62
Oct 24 11:00:38 filecoin lotus-worker[20925]:    7: std::sys_common::backtrace::print
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libstd/sys_common/backtrace.rs:49
Oct 24 11:00:38 filecoin lotus-worker[20925]:    8: std::panicking::default_hook::{{closure}}
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libstd/panicking.rs:204
Oct 24 11:00:38 filecoin lotus-worker[20925]:    9: std::panicking::default_hook
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libstd/panicking.rs:224
Oct 24 11:00:38 filecoin lotus-worker[20925]:   10: std::panicking::rust_panic_with_hook
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libstd/panicking.rs:470
Oct 24 11:00:38 filecoin lotus-worker[20925]:   11: rust_begin_unwind
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libstd/panicking.rs:378
Oct 24 11:00:38 filecoin lotus-worker[20925]:   12: core::panicking::panic_fmt
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libcore/panicking.rs:85
Oct 24 11:00:38 filecoin lotus-worker[20925]:   13: core::option::expect_none_failed
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libcore/option.rs:1211
Oct 24 11:00:38 filecoin lotus-worker[20925]:   14: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
Oct 24 11:00:38 filecoin lotus-worker[20925]:   15: std::panicking::try::do_call
Oct 24 11:00:38 filecoin lotus-worker[20925]:   16: __rust_maybe_catch_panic
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libpanic_unwind/lib.rs:86
Oct 24 11:00:38 filecoin lotus-worker[20925]:   17: <rayon_core::job::HeapJob<BODY> as rayon_core::job::Job>::execute
Oct 24 11:00:38 filecoin lotus-worker[20925]:   18: rayon_core::registry::WorkerThread::wait_until_cold
Oct 24 11:00:38 filecoin lotus-worker[20925]:   19: rayon_core::scope::ScopeBase::steal_till_jobs_complete
Oct 24 11:00:38 filecoin lotus-worker[20925]:   20: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
Oct 24 11:00:38 filecoin lotus-worker[20925]:   21: std::panicking::try::do_call
Oct 24 11:00:38 filecoin lotus-worker[20925]:   22: __rust_maybe_catch_panic
Oct 24 11:00:38 filecoin lotus-worker[20925]:              at src/libpanic_unwind/lib.rs:86
Oct 24 11:00:38 filecoin lotus-worker[20925]:   23: <rayon_core::job::StackJob<L,F,R> as rayon_core::job::Job>::execute
Oct 24 11:00:38 filecoin lotus-worker[20925]:   24: rayon_core::registry::WorkerThread::wait_until_cold
Oct 24 11:00:38 filecoin lotus-worker[20925]:   25: rayon_core::registry::ThreadBuilder::run
```

This will loop 3 times until lotus demotes the sector back into PC1. Then it will go through PC1 again and then PC2 successfully. Seems either aborting PC2 corrupts something in the generated PC1 files, or something is leaked out onto the file system which prevents PC2 from running again on the same files.

**Comment by karalabe** [comment]
Hmmm, this one actually managed to resume when doing the second try of PC2.

The first try was hung for 2 hours, never reached even printing `persisting base tree_c 1/8 [...]`, then errored  out with the bad file descriptor. After lotus restarting it, it's at `persisting base tree_c 2/8` already after 10 minutes.

**Comment by karalabe** [comment]
No, unfortunately that quick thing I mentioned above was a different PC2 on the same worker. This one keeps choking with bad file descriptors in a loop.

**Comment by karalabe** [comment]
```
6.	2020-10-24 09:17:34 +0300 EEST:	[event;sealing.SectorRestart]	{"User":{}}
7.	2020-10-24 11:00:53 +0300 EEST:	[event;sealing.SectorSealPreCommit2Failed]	{"User":{}}
	seal pre commit(2) failed: presealing sector 357 (): Rust panic: no unwind information
8.	2020-10-24 11:01:53 +0300 EEST:	[event;sealing.SectorRetrySealPreCommit2]	{"User":{}}
9.	2020-10-24 11:29:02 +0300 EEST:	[event;sealing.SectorSealPreCommit2Failed]	{"User":{}}
	seal pre commit(2) failed: presealing sector 357 (): Rust panic: no unwind information
10.	2020-10-24 11:30:02 +0300 EEST:	[event;sealing.SectorRetrySealPreCommit2]	{"User":{}}
```

Almost 2 hours for the first crash. 28 minutes to the second.

**Comment by karalabe** [comment]
This is getting super annoying. An RPC call timed out between the miner-worker, all tasks were dropped, then when PC2s were removed, they all fail with "Bad file descriptor" until the sectors get nuked back into PC1. Every day I lose 3-4 PC1 runs because PC2 fails for whatever reason and takes the sector with it.

**Comment by porcuquine** [comment]
@cryptonemo Can you look into this and/or do you have immediate insights on the cause?

**Comment by cryptonemo** [comment]
Make sure `ulimit -n` reports something higher than the default (e.g. 20000).

**Comment by karalabe** [comment]
It's set to 64K

**Comment by cryptonemo** [comment]
Try upping it then, or running less in parallel?  I've only seen this error when file descriptor limits are hit.

**Comment by karalabe** [comment]
Doesn't matter. Today I had to restart my miner, it did 1 single PC2 and nothing else, failed with bad file descriptors. Something's rotting in the file handling.

**Comment by karalabe** [comment]
```
Oct 31 09:58:22 filecoin lotus-worker[792]: thread '<unnamed>' panicked at 'failed to flatten and write store: Bad file descriptor (os error 9)
Oct 31 09:58:22 filecoin lotus-worker[792]: Stack backtrace:
Oct 31 09:58:22 filecoin lotus-worker[792]:    0: <merkletree::store::disk::DiskStore<E> as merkletree::store::Store<E>>::copy_from_slice
Oct 31 09:58:22 filecoin lotus-worker[792]:    1: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
Oct 31 09:58:22 filecoin lotus-worker[792]:    2: rayon::iter::plumbing::Folder::consume_iter
Oct 31 09:58:22 filecoin lotus-worker[792]:    3: rayon::iter::plumbing::bridge_producer_consumer::helper
Oct 31 09:58:22 filecoin lotus-worker[792]:    4: rayon_core::join::join_context::{{closure}}
Oct 31 09:58:22 filecoin lotus-worker[792]:    5: rayon_core::join::join_context
Oct 31 09:58:22 filecoin lotus-worker[792]:    6: rayon::iter::plumbing::bridge_producer_consumer::helper
Oct 31 09:58:22 filecoin lotus-worker[792]:    7: rayon_core::join::join_context::{{closure}}
Oct 31 09:58:22 filecoin lotus-worker[792]:    8: rayon_core::join::join_context
Oct 31 09:58:22 filecoin lotus-worker[792]:    9: rayon::iter::plumbing::bridge_producer_consumer::helper
Oct 31 09:58:22 filecoin lotus-worker[792]:   10: rayon_core::join::join_context::{{closure}}
Oct 31 09:58:22 filecoin lotus-worker[792]:   11: rayon_core::join::join_context
Oct 31 09:58:22 filecoin lotus-worker[792]:   12: rayon::iter::plumbing::bridge_producer_consumer::helper
Oct 31 09:58:22 filecoin lotus-worker[792]:   13: rayon_core::join::join_context::{{closure}}
Oct 31 09:58:22 filecoin lotus-worker[792]:   14: rayon_core::join::join_context
Oct 31 09:58:22 filecoin lotus-worker[792]:   15: rayon::iter::plumbing::bridge_producer_consumer::helper
Oct 31 09:58:22 filecoin lotus-worker[792]:   16: rayon_core::join::join_context::{{closure}}
Oct 31 09:58:22 filecoin lotus-worker[792]:   17: rayon_core::join::join_context
Oct 31 09:58:22 filecoin lotus-worker[792]:   18: rayon::iter::plumbing::bridge_producer_consumer::helper
Oct 31 09:58:22 filecoin lotus-worker[792]:   19: rayon_core::join::join_context::{{closure}}
Oct 31 09:58:22 filecoin lotus-worker[792]:   20: rayon_core::join::join_context
Oct 31 09:58:22 filecoin lotus-worker[792]:   21: rayon::iter::plumbing::bridge_producer_consumer::helper
Oct 31 09:58:22 filecoin lotus-worker[792]:   22: rayon::iter::plumbing::bridge
Oct 31 09:58:22 filecoin lotus-worker[792]:   23: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
Oct 31 09:58:22 filecoin lotus-worker[792]:   24: std::panicking::try
Oct 31 09:58:22 filecoin lotus-worker[792]:   25: <rayon_core::job::HeapJob<BODY> as rayon_core::job::Job>::execute
Oct 31 09:58:22 filecoin lotus-worker[792]:   26: rayon_core::registry::WorkerThread::wait_until_cold
Oct 31 09:58:22 filecoin lotus-worker[792]:   27: rayon_core::scope::ScopeBase::steal_till_jobs_complete
Oct 31 09:58:22 filecoin lotus-worker[792]:   28: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
Oct 31 09:58:22 filecoin lotus-worker[792]:   29: <rayon_core::job::StackJob<L,F,R> as rayon_core::job::Job>::execute
Oct 31 09:58:22 filecoin lotus-worker[792]:   30: rayon_core::registry::WorkerThread::wait_until_cold
Oct 31 09:58:22 filecoin lotus-worker[792]:   31: rayon_core::registry::ThreadBuilder::run
Oct 31 09:58:22 filecoin lotus-worker[792]:   32: std::sys_common::backtrace::__rust_begin_short_backtrace
Oct 31 09:58:22 filecoin lotus-worker[792]:   33: core::ops::function::FnOnce::call_once{{vtable.shim}}
Oct 31 09:58:22 filecoin lotus-worker[792]:   34: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
Oct 31 09:58:22 filecoin lotus-worker[792]:              at ./rustc/beb5ae474d2835962ebdf7416bd1c9ad864fe101/library/alloc/src/boxed.rs:1042
Oct 31 09:58:22 filecoin lotus-worker[792]:       <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
Oct 31 09:58:22 filecoin lotus-worker[792]:              at ./rustc/beb5ae474d2835962ebdf7416bd1c9ad864fe101/library/alloc/src/boxed.rs:1042
Oct 31 09:58:22 filecoin lotus-worker[792]:       std::sys::unix::thread::Thread::new::thread_start
Oct 31 09:58:22 filecoin lotus-worker[792]:              at ./rustc/beb5ae474d2835962ebdf7416bd1c9ad864fe101/library/std/src/sys/unix/thread.rs:87
Oct 31 09:58:22 filecoin lotus-worker[792]:   35: start_thread
Oct 31 09:58:22 filecoin lotus-worker[792]:   36: clone', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:601:64
Oct 31 09:58:22 filecoin lotus-worker[792]: stack backtrace:
Oct 31 09:58:22 filecoin lotus-worker[792]:    0: rust_begin_unwind
Oct 31 09:58:22 filecoin lotus-worker[792]:              at ./rustc/beb5ae474d2835962ebdf7416bd1c9ad864fe101/library/std/src/panicking.rs:483
Oct 31 09:58:22 filecoin lotus-worker[792]:    1: core::panicking::panic_fmt
Oct 31 09:58:22 filecoin lotus-worker[792]:              at ./rustc/beb5ae474d2835962ebdf7416bd1c9ad864fe101/library/core/src/panicking.rs:85
Oct 31 09:58:22 filecoin lotus-worker[792]:    2: core::option::expect_none_failed
Oct 31 09:58:22 filecoin lotus-worker[792]:              at ./rustc/beb5ae474d2835962ebdf7416bd1c9ad864fe101/library/core/src/option.rs:1234
Oct 31 09:58:22 filecoin lotus-worker[792]:    3: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
Oct 31 09:58:22 filecoin lotus-worker[792]:    4: std::panicking::try
Oct 31 09:58:22 filecoin lotus-worker[792]:    5: <rayon_core::job::HeapJob<BODY> as rayon_core::job::Job>::execute
Oct 31 09:58:22 filecoin lotus-worker[792]:    6: rayon_core::registry::WorkerThread::wait_until_cold
Oct 31 09:58:22 filecoin lotus-worker[792]:    7: rayon_core::scope::ScopeBase::steal_till_jobs_complete
Oct 31 09:58:22 filecoin lotus-worker[792]:    8: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
Oct 31 09:58:22 filecoin lotus-worker[792]:    9: <rayon_core::job::StackJob<L,F,R> as rayon_core::job::Job>::execute
Oct 31 09:58:22 filecoin lotus-worker[792]:   10: rayon_core::registry::WorkerThread::wait_until_cold
Oct 31 09:58:22 filecoin lotus-worker[792]:   11: rayon_core::registry::ThreadBuilder::run
```

**Comment by cryptonemo** [comment]
@karalabe Can you try running again with `FIL_PROOFS_COLUMN_WRITE_BATCH_SIZE=16384`?

**Comment by cryptonemo** [comment]
This is also related to #1199

**Comment by jennijuju** [comment]
A miner thia is on lotus v1.2.1 had the same issue 
with `FIL_PROOFS_COLUMN_WRITE_BATCH_SIZE =262144` and ulimit=1048576

**Comment by cryptonemo** [comment]
The solution is to raise the file descriptor limits.  The available descriptor count is now logged with `RUST_LOG=info`.  If the value is as high as expected and the failure still occurs, please open a new issue with the specifics.
